// @keystone-alpha/fields/dist/chunk-8faa95af.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var core = require('@emotion/core');

var React = require('react');

var React__default = _interopDefault(React);

var fields = require('@arch-ui/fields');

var _extends = _interopDefault(require("@babel/runtime/helpers/extends"));

var button = require('@arch-ui/button');

var Popout = _interopDefault(require('@arch-ui/popout'));

var SketchPicker = _interopDefault(require('react-color/lib/Sketch'));

const ColorField = _ref => {
  let field = _ref.field,
      serverValue = _ref.value,
      errors = _ref.errors,
      _onChange = _ref.onChange;
  const value = serverValue || '';
  const htmlID = "ks-input-".concat(field.path);

  const target = props => core.jsx(button.Button, _extends({}, props, {
    variant: "ghost"
  }), value ? core.jsx(React__default.Fragment, null, core.jsx("div", {
    style: {
      // using inline styles instead of emotion for setting the color
      // since emotion doesn't escape styles so it could be used for CSS injection
      // this is also better in terms of memory since the value can change a lot
      // and emotion caches everything
      backgroundColor: value
    },
    css: {
      borderRadius: 3,
      display: 'inline-block',
      height: 18,
      width: 18,
      marginRight: 10,
      verticalAlign: 'middle'
    }
  }), core.jsx("span", {
    css: {
      verticalAlign: 'middle'
    }
  }, value)) : 'Set Color');

  return core.jsx(fields.FieldContainer, null, core.jsx(fields.FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), core.jsx(fields.FieldInput, null, core.jsx(Popout, {
    width: 220,
    target: target
  }, core.jsx(SketchPicker, {
    css: {
      // using !important because react-color uses inline styles and applies a box shadow
      // but Popout already applies a box shadow
      boxShadow: 'none !important'
    },
    noAlpha: true,
    color: value,
    onChange: _ref2 => {
      let hex = _ref2.hex;

      _onChange(hex);
    }
  }))));
};

exports.default = ColorField;
