// @keystone-alpha/fields/dist/chunk-f41cd08b.esm.js
import { jsx } from '@emotion/core';
import React__default from 'react';
import { FieldContainer, FieldLabel, FieldInput } from '@arch-ui/fields';
import _extends from '@babel/runtime/helpers/esm/extends';
import { Button } from '@arch-ui/button';
import Popout from '@arch-ui/popout';
import SketchPicker from 'react-color/lib/Sketch';

const ColorField = _ref => {
  let field = _ref.field,
      serverValue = _ref.value,
      errors = _ref.errors,
      _onChange = _ref.onChange;
  const value = serverValue || '';
  const htmlID = "ks-input-".concat(field.path);

  const target = props => jsx(Button, _extends({}, props, {
    variant: "ghost"
  }), value ? jsx(React__default.Fragment, null, jsx("div", {
    style: {
      // using inline styles instead of emotion for setting the color
      // since emotion doesn't escape styles so it could be used for CSS injection
      // this is also better in terms of memory since the value can change a lot
      // and emotion caches everything
      backgroundColor: value
    },
    css: {
      borderRadius: 3,
      display: 'inline-block',
      height: 18,
      width: 18,
      marginRight: 10,
      verticalAlign: 'middle'
    }
  }), jsx("span", {
    css: {
      verticalAlign: 'middle'
    }
  }, value)) : 'Set Color');

  return jsx(FieldContainer, null, jsx(FieldLabel, {
    htmlFor: htmlID,
    field: field,
    errors: errors
  }), jsx(FieldInput, null, jsx(Popout, {
    width: 220,
    target: target
  }, jsx(SketchPicker, {
    css: {
      // using !important because react-color uses inline styles and applies a box shadow
      // but Popout already applies a box shadow
      boxShadow: 'none !important'
    },
    noAlpha: true,
    color: value,
    onChange: _ref2 => {
      let hex = _ref2.hex;

      _onChange(hex);
    }
  }))));
};

export default ColorField;
