// @keystone-alpha/field-content/dist/chunk-0bf1bb91.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

require("@babel/runtime/helpers/objectWithoutProperties");

var _extends = _interopDefault(require("@babel/runtime/helpers/extends"));

var core = require('@emotion/core');

var __chunk_2 = require('./chunk-ea748c32.cjs.js');

var icons = require('@arch-ui/icons');

require('@arch-ui/theme');

require('@arch-ui/color-utils');

require('@arch-ui/tooltip');

require('@arch-ui/typography');

var __chunk_3 = require('./chunk-6384d08f.cjs.js');

let type = 'blockquote';

function ToolbarElement(_ref) {
  let editor = _ref.editor,
      editorState = _ref.editorState;

  let hasBlockquote = __chunk_2.hasAncestorBlock(editorState, type);

  return core.jsx(__chunk_3.ToolbarButton, {
    isActive: hasBlockquote,
    icon: core.jsx(icons.QuoteIcon, null),
    label: "Blockquote",
    onClick: () => {
      if (hasBlockquote) {
        editor.unwrapBlock(type);
      } else {
        editor.wrapBlock(type);
      }

      editor.focus();
    }
  });
}

function Node(_ref2) {
  let attributes = _ref2.attributes,
      children = _ref2.children;
  return core.jsx("blockquote", _extends({}, attributes, {
    css: {
      borderLeft: '4px solid #eee',
      color: '#666',
      fontStyle: 'italic',
      margin: 0,
      marginBottom: '1em',
      paddingLeft: '1em'
    }
  }), children);
}

exports.type = type;
exports.ToolbarElement = ToolbarElement;
exports.Node = Node;
