// @keystone-alpha/field-content/dist/chunk-3c8b2136.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var _slicedToArray = _interopDefault(require("@babel/runtime/helpers/slicedToArray"));

var React = require('react');

var React__default = _interopDefault(React);

function useScrollListener(handler) {
  React.useEffect(() => {
    window.addEventListener('scroll', handler);
    return () => window.removeEventListener('scroll', handler);
  }, [handler]);
}

let isBrowser = typeof window !== 'undefined';

function getSize() {
  if (isBrowser) {
    return {
      innerHeight: window.innerHeight,
      innerWidth: window.innerWidth
    };
  }

  return {
    innerHeight: 0,
    innerWidth: 0
  };
}

function useWindowSize() {
  let _useState = React.useState(getSize()),
      _useState2 = _slicedToArray(_useState, 2),
      windowSize = _useState2[0],
      setWindowSize = _useState2[1];

  React.useEffect(() => {
    function handleResize() {
      setWindowSize(getSize());
    }

    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, [setWindowSize]);
  return windowSize;
}

function useStateWithEqualityCheck(initialValue) {
  let _useState = React.useState(initialValue),
      _useState2 = _slicedToArray(_useState, 2),
      value = _useState2[0],
      _setValue = _useState2[1];

  let setValue = React.useCallback(newValue => {
    if (newValue !== value) {
      _setValue(newValue);
    }
  }, [value]);
  return [value, setValue];
}

exports.useStateWithEqualityCheck = useStateWithEqualityCheck;
exports.useWindowSize = useWindowSize;
exports.useScrollListener = useScrollListener;
