// @keystone-alpha/field-content/dist/chunk-28658b2d.esm.js
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import '@babel/runtime/helpers/esm/extends';
import '@emotion/core';
import { b as hasBlock } from './chunk-ec3635e1.esm.js';
import '@arch-ui/theme';
import '@arch-ui/color-utils';
import '@arch-ui/tooltip';
import '@arch-ui/typography';
import { a as ToolbarButton } from './chunk-31fbf0c2.esm.js';
import { createElement } from 'react';
import { type } from './chunk-be6a2a7f.esm.js';
let type$1 = 'heading';

function ToolbarElement(_ref) {
  let editor = _ref.editor,
      editorState = _ref.editorState;
  return createElement(ToolbarButton, {
    icon: createElement("span", {
      "aria-hidden": true
    }, "H"),
    label: "Heading",
    isActive: hasBlock(editorState, type$1),
    onClick: () => {
      if (hasBlock(editorState, type$1)) {
        editor.setBlocks({
          type: type
        });
      } else {
        editor.setBlocks({
          type: type$1
        });
      }

      editor.focus();
    }
  });
}

function Node(_ref2) {
  let attributes = _ref2.attributes,
      children = _ref2.children;
  return createElement("h2", attributes, children);
}

let getPlugins = () => [{
  onKeyDown(event, editor, next) {
    // make it so when you press enter after typing a heading,
    // the block type will change to a paragraph
    if (event.keyCode === 13 && editor.value.blocks.every(block => block.type === type$1)) {
      editor.splitBlock().setBlocks(type);
      return;
    }

    next();
  }

}];

export { type$1 as type, ToolbarElement, Node, getPlugins };
