// @keystone-alpha/field-content/dist/chunk-817d404d.esm.js
import '@babel/runtime/helpers/esm/objectWithoutProperties';
import _extends from '@babel/runtime/helpers/esm/extends';
import { jsx } from '@emotion/core';
import { a as hasAncestorBlock } from './chunk-ec3635e1.esm.js';
import { QuoteIcon } from '@arch-ui/icons';
import '@arch-ui/theme';
import '@arch-ui/color-utils';
import '@arch-ui/tooltip';
import '@arch-ui/typography';
import { a as ToolbarButton } from './chunk-31fbf0c2.esm.js';
let type = 'blockquote';

function ToolbarElement(_ref) {
  let editor = _ref.editor,
      editorState = _ref.editorState;
  let hasBlockquote = hasAncestorBlock(editorState, type);
  return jsx(ToolbarButton, {
    isActive: hasBlockquote,
    icon: jsx(QuoteIcon, null),
    label: "Blockquote",
    onClick: () => {
      if (hasBlockquote) {
        editor.unwrapBlock(type);
      } else {
        editor.wrapBlock(type);
      }

      editor.focus();
    }
  });
}

function Node(_ref2) {
  let attributes = _ref2.attributes,
      children = _ref2.children;
  return jsx("blockquote", _extends({}, attributes, {
    css: {
      borderLeft: '4px solid #eee',
      color: '#666',
      fontStyle: 'italic',
      margin: 0,
      marginBottom: '1em',
      paddingLeft: '1em'
    }
  }), children);
}

export { type, ToolbarElement, Node };
