// @keystone-alpha/field-content/dist/chunk-78fcb4df.cjs.js
'use strict';

function _interopDefault(ex) {
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

require("@babel/runtime/helpers/objectWithoutProperties");

require("@babel/runtime/helpers/extends");

require('@emotion/core');

var __chunk_2 = require('./chunk-ea748c32.cjs.js');

require('@arch-ui/theme');

require('@arch-ui/color-utils');

require('@arch-ui/tooltip');

require('@arch-ui/typography');

var __chunk_3 = require('./chunk-6384d08f.cjs.js');

var React = require('react');

var React__default = _interopDefault(React);

var paragraph_js = require('./chunk-b150bff8.cjs.js');

let type = 'heading';

function ToolbarElement(_ref) {
  let editor = _ref.editor,
      editorState = _ref.editorState;
  return React.createElement(__chunk_3.ToolbarButton, {
    icon: React.createElement("span", {
      "aria-hidden": true
    }, "H"),
    label: "Heading",
    isActive: __chunk_2.hasBlock(editorState, type),
    onClick: () => {
      if (__chunk_2.hasBlock(editorState, type)) {
        editor.setBlocks({
          type: paragraph_js.type
        });
      } else {
        editor.setBlocks({
          type: type
        });
      }

      editor.focus();
    }
  });
}

function Node(_ref2) {
  let attributes = _ref2.attributes,
      children = _ref2.children;
  return React.createElement("h2", attributes, children);
}

let getPlugins = () => [{
  onKeyDown(event, editor, next) {
    // make it so when you press enter after typing a heading,
    // the block type will change to a paragraph
    if (event.keyCode === 13 && editor.value.blocks.every(block => block.type === type)) {
      editor.splitBlock().setBlocks(paragraph_js.type);
      return;
    }

    next();
  }

}];

exports.type = type;
exports.ToolbarElement = ToolbarElement;
exports.Node = Node;
exports.getPlugins = getPlugins;
